feat: add a use cases template for github issues

laurensdeb opened the pull request:
This is a proposal for a Github issues template for use cases for the LWS WG, all feedback is welcome.


The template was loosely inspired on
[VC Use Cases](https://www.w3.org/TR/vc-use-cases/)
[DID Use Cases](https://www.w3.org/TR/did-use-cases/)
[WOT Use Cases](https://www.w3.org/TR/wot-usecases/#digital-twin-2)

laurensdeb committed:
feat: add a use cases template for github issues

acoburn reviewed (approved):
LGTM

renyuneyun commented:
First of all, great to see this moving. More than happy to convert my existing ones following a template.
But I do have questions (see below) on the template itself, from a newbie's perspective. Hope more explanations or examples are added, or template being updated.

1. Does "precondition" refer to preconditions in terms of features or other use cases, or precondition of the exact use case (e.g. what types of data are stored in the Pod, and then the use case exists)?
2. What does "trigger" really mean? Why and how is it different from the scenario itself?
3. What should be included in "alternative cases"? In particular, do they refer to "alternative use cases (which may fulfil the same / comparable functionality of this use case)"?

csarven commented on .github/ISSUE_TEMPLATE/use-case.md:
```diff
@@ -0,0 +1,44 @@
+
+## Preconditions:
+
+*What preconditions do we expect there to be fulfilled prior to this use case being triggered?*
```
```suggestion
*What conditions must be in place or assumed before this use case can begin?*
```

csarven commented on .github/ISSUE_TEMPLATE/use-case.md:
```diff
@@ -0,0 +1,44 @@
+
+## Trigger:
+
+*What triggers this use case?*
```
The term "trigger" is not particularly clear - as also mentioned in https://github.com/w3c/lws-ucs/pull/5#issuecomment-2444539553 . Perhaps what's intended is along the following lines?

```suggestion
*What (user or system) event or action initiates this use case?*
```

csarven commented on .github/ISSUE_TEMPLATE/use-case.md:
```diff
@@ -0,0 +1,44 @@
+
+## Distinction:
+
+*What is the challenge in realizing this use case? What distinguishes it from other, similar use cases?*
```
My reading of "challenge in realizing" that it is either unclear here or works with some knowledge of potential solutions or short-comings of solutions. Better to connect it with distinguishing factors

```suggestion
*What unique challenges or distinguishing factors, like technical issues, user experience needs, or workflow integration, are associated with this use case?*
```

Alternatively drop the ", like..integration, " above.

TallTed replied:
```suggestion
*What unique challenges or distinguishing factors (like technical issues, user experience needs, workflow integration, etc.) are associated with this use case?*
```

csarven commented on .github/ISSUE_TEMPLATE/use-case.md:
```diff
@@ -0,0 +1,44 @@
+
+## Scenario:
+
+*A more detailed happy-case scenario where this use case would be relevant*
```
```suggestion
*Describe an ideal or happy-case scenario where this use case would play out as intended.*
```

csarven commented on .github/ISSUE_TEMPLATE/use-case.md:
```diff
@@ -0,0 +1,44 @@
+
+## Alternative case(s):
+
+*What alternative flows do we think a system must cover for this use case?*
```
Emphasis on different (from happy-case) but still a valid route:

```suggestion
*What alternative flows or variations should the system handle for this use case?*
```

csarven commented on .github/ISSUE_TEMPLATE/use-case.md:
```diff
@@ -0,0 +1,44 @@
+
+## Acceptance Criteria:
+
+*When do we consider this use case to be sufficiently covered by the specification, what limitations do we consider acceptable?*
```
```suggestion
*When do we consider this use case to be sufficiently covered by the specification, what limitations do we consider acceptable?*
```

csarven commented on .github/ISSUE_TEMPLATE/use-case.md:
```diff
@@ -0,0 +1,44 @@
+## Alternative case(s):
+
+*What alternative flows do we think a system must cover for this use case?*
+
```
Happy, alternative, and error scenarios serve different purposes:

```suggestion

## Error scenario

*What unexpected issues or errors might arise, and how should the system handle them?*

```

laurensdeb replied:
For me an error scenario could be seen as an alternative scenario, but happy to make this distinction if it can provide additional clarity for the use case description.

csarven replied:
I agree that the error scenario can be seen as an alternative scenario. I don't have a strong opinion on it but I do think the distinction is worthwhile. At the end, if this information can be explicitly described, it can be considered. Otherwise, it may or may not be taken into account when requirements are proposed and reviewed, and possibly codified. It will slip through the specification cracks ;)

The distinction I tried to make with happy and alternative was that they are successful - the latter being a legitimate path towards the goal - and well error is simply a fail.

strawperson use case: user tries to save their article in their storage. storage says nope because they don't have enough disk space. what an application/user to do?

(so, an alternative could hypothetically be that the user contacts their admin by email or ssh into their server to free up space or whatever, but that doesn't sound exciting. so what would the system need to do to handle the error and still keep the user (and their application for instance) in their flow, and continue - instructions to the app to inform the client and/or try quietly behind the scenes? have their app catch fire? send a structured message to the client ...

another: user tries to access a resource and server says that they're unauthorized. what happens next? we could tell the client that "we know you're unauthorized but if you really think you should have access to this, maybe you can contact us.." (see for instance client errors: https://github.com/solid/specification/pull/253#issuecomment-895062229 -- these are actual needs and implementation).

so, while happy/alternative are fine, but error to me says something a bit different and important.

renyuneyun replied:
I'm confused by what "alternative use case" means, and slightly "error use case" (maybe still due to my lack of experience). Based on all the comments above, my understanding is like this:

- "Happy use case": The primary use case / flow of user interaction for making the use case happen, leading to an expected outcome;
- "Alternative use case": Other user interactions that will result in (?, or trigger) the same (?, or similar, in terms of the technical details of the expected outcome) use case / expected outcome (?, or technical details), in the same (?, or another) app;
- "Error use case": When executing the use case / interaction, what errors / exceptions may emerge, and (if known) how (?, or criteria on how) to deal with them.

Every question mark indicates a confusion, to the word / phrase before it; they also include a description of the alternatives (thus confusions) I see.

csarven commented on .github/ISSUE_TEMPLATE/use-case.md:
```diff
@@ -0,0 +1,44 @@
+
+## References:
+
+*Any relevant references, for example to practical examples of this use case or possible solutions (in other domains)*
```
```suggestion
*List any relevant resources or examples that could inform this use case, possibly from other domains or solutions.*
```

csarven commented on .github/ISSUE_TEMPLATE/use-case.md:
```diff
@@ -0,0 +1,44 @@
+about: Use this template to propose an LWS use case.
+title: "[UC] <<brief description of use case>>"
+labels: triage, usecase
+assignees: ''
```
I'm not sure I understand this in the template.

```suggestion
assignees: ''
```

csarven reviewed (changes requested):
Added some clarifications and suggestions for group's consideration.

Perhaps instead of or in addition to "assignee", the proposer should indicate things along these lines. So this could be integrate into the template:

* [ ] will implement potential solutions
* [ ] will share implementation experience
* [ ] will share code implementing the requirements that come out of this use case

chrisn commented:
LGTM, and support csarven's suggestions

laurensdeb committed:
feat: apply suggestions from review

Co-authored-by: Sarven Capadisli <info@csarven.ca>
Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

laurensdeb committed:
feat: apply suggestions from review

Co-authored-by: Sarven Capadisli <info@csarven.ca>

laurensdeb committed:
chore: remove assignees

laurensdeb commented:
@renyuneyun This is still a first revision of the use case template, I'm sure when actually using it for the concrete use cases it will still need some refinement.

To answer your questions:

> Does "precondition" refer to preconditions in terms of features or other use cases, or precondition of the exact use case 
> (e.g. what types of data are stored in the Pod, and then the use case exists)?

I would interpret preconditions quite broadly, these could also refer to other use cases. For example, "a user must have a LWS-compliant storage server" or "a resource must have been created in the storage server".

> What does "trigger" really mean? Why and how is it different from the scenario itself?

Typically I see a use case as something which is triggered by a particular event. For example, a user stores a resource in a LWS-compliant storage server such that it can be shared or retrieved later on (the use case) by performing an HTTP request to the storage server (a possible trigger for the use case, of which there can be several).

> What should be included in "alternative cases"? In particular, do they refer to "alternative use cases (which may fulfil the
> same / comparable functionality of this use case)"?

Alternative cases are instances of a use case which differ from the "happy case" scenario in a significant way, because some exception has occured during the happy scenario for example. For example, an IdP is unavailable for the storage server when trying to authenticate a client, how might we handle such a scenario.

laurensdeb commented:
@csarven I've updated the pull request to reflect your proposed changes.

@hzbarcea I've understood from @acoburn that you would be finalizing the template, so I've added you as assignee. If you need any further input from my side, let me know. Else, I will leave this PR for you to merge or close.

renyuneyun commented on .github/ISSUE_TEMPLATE/use-case.md:
```diff
@@ -0,0 +1,43 @@
+labels: triage, usecase
+
+---
+
```
Maybe add a clarification for the mandatory status for each section (either here or at the section heading)? Or are all fields mandatory for the first draft?

hzbarcea committed:
Update .github/ISSUE_TEMPLATE/use-case.md

Co-authored-by: Sarven Capadisli <info@csarven.ca>
